nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKudu.java

```diff
@@ -0,0 +1,191 @@
+            .name("KUDU Masters")
+            .description("List all kudu masters's ip with port (e.g. 7051), comma separated")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    protected static final PropertyDescriptor TABLE_NAME = new PropertyDescriptor.Builder()
+            .name("Table Name")
+            .description("The name of the Kudu Table to put data into")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor RECORD_READER = new PropertyDescriptor.Builder()
+            .name("record-reader")
+            .displayName("Record Reader")
+            .description("The service for reading records from incoming flow files.")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(true)
+            .build();
+
+    protected static final PropertyDescriptor SKIP_HEAD_LINE = new PropertyDescriptor.Builder()
+            .name("Skip head line")
+            .description("Set it to true if your first line is the header line e.g. column names")
+            .allowableValues("true", "false")
+            .defaultValue("true")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    protected static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("A FlowFile is routed to this relationship after it has been successfully stored in Kudu")
+            .build();
+    protected static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("A FlowFile is routed to this relationship if it cannot be sent to Kudu")
+            .build();
+
+    public static final String RECORD_COUNT_ATTR = "record.count";
+
+    protected String kuduMasters;
+    protected String tableName;
+    protected boolean skipHeadLine;
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        tableName = context.getProperty(TABLE_NAME).getValue();
+        kuduMasters = context.getProperty(KUDU_MASTERS).getValue();
+        skipHeadLine = context.getProperty(SKIP_HEAD_LINE).asBoolean();
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        try {
+            if (flowFile == null) return;
+            final Map<String,String> attributes = new HashMap<String, String>();
+            final AtomicReference<Throwable> exceptionHolder = new AtomicReference<>(null);
+            final RecordReaderFactory recordReaderFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+            final KuduClient kuduClient = getKuduConnection(kuduMasters);
```

Comment:
He had it in on enabled previously. I encouraged moving to a lazy creation model because otherwise if the connection breaks there was no logic to restore it without manual intervention.  It should not happen per flow file either though of course.  Some sort of connection pooling would be appropriate.  I dont know enough about the kudu connection client logic to have better advice there but onEnabled is not great and per flowfile is probably not either.